openhtf/__init__.py

```diff
@@ -84,32 +84,56 @@ def __call__(self, test_record):  # pylint: disable=invalid-name
       f.write(self.encode(as_dict))
 
 
-def TestPhase(timeout_s=None, run_if=None):  # pylint: disable=invalid-name
-  """Decorator to wrap a test phase function with the given options.
-
-  Args:
+class TestPhaseOptions(mutablerecords.Record(
+    'TestPhaseOptions', [], {'timeout_s': None, 'run_if': None})):
+  """
+  Attributes:
     timeout_s: Timeout to use for the phase, in seconds.
     run_if: Callback that decides whether to run the phase or not.  The
       callback will be passed the phase_data the phase would be run with.
 
-  Returns:
-    A wrapper function that takes a phase_func and returns a
-        TestPhaseInfo for it with the given options set.
+  Example Usage:
+    @TestPhaseOptions(timeout_s=1)
+    def PhaseFunc(test):
+      pass
   """
 
-  def Wrap(phase_func):  # pylint: disable=invalid-name
-    """Attach the given options to the phase_func."""
+  def __call__(self, phase_func):
+    phase = TestPhaseInfo.WrapOrReturn(phase_func)
+    phase.options = self
+    return phase
 
-    # Test Phases must take at least one argument (the phase data tuple).
-    if len(inspect.getargspec(phase_func).args) < 1:
-      raise InvalidTestPhaseError(phase_func, 'Not enough args')
+PhasePlug = mutablerecords.Record('PhasePlug', ['name', 'cls', 'update_kwargs'])
 
-    if timeout_s is not None:
-      phase_func.timeout_s = timeout_s
-    if run_if is not None:
-      phase_func.run_if = run_if
-    return phase_func
-  return Wrap
+
+class TestPhaseInfo(mutablerecords.Record(
```

Comment:
I'll defer to Joe on this one, but we had made the call a while back that we wanted phase functions to remain just functions.  In fact, this was explicitly changed from the way it was done in XTF (which is what you've recreated here).

Since then, we ended up tacking a few attributes onto there, so arguably it really should be a fancier class like this.  I'm somewhat 50/50 on it, so like I said I'll let Joe weigh in.

Again, though, this was an obvious design decision that was made and explicitly changed from XTF, but I didn't see any conversation around "hey guys why did you do this?" or "i think we should change this, can I go ahead and implement it?"